### PR TITLE
doc: add write/read example for interact doc

### PR DESCRIPTION
**Pull request by xiang90**
/cc @joshix @heyitsanthony

### CODE REVIEW

File: Documentation/interact_v3.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Interact with etcd
```
Comment (philips) [inline Documentation/interact_v3.md]: Interacting instead of interact

File: Documentation/interact_v3.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+bar_new
+
+# watch for changes on key `foo` since revision 3
+$ etcdctl watch --rev=3 foo
+PUT
+foo
+bar_new
+```
+
+## Compact revisions
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Compacted revisions

File: Documentation/interact_v3.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+As we mentioned, etcd keeps revisions so that applications can read past versions of keys. However, applications may want to compact revisions to avoid unbounded data growing. After a compaction, etcd removes historical revisions and release resources for future use. All revisions before the compacted revision will be unavailable for accessing.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: etcd keeps revisions so that applications can read past versions of keys. However, to avoid accumulating an unbounded amount of history, it is important to compact away past revisions. After a compaction, etcd removes historical revisions, releasing resources for future use. All superseded data with revisions before the compacted revision will be unavailable.

File: Documentation/interact_v3.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+foo
+bar
+# in another terminal: etcdctl put foo1 bar1
+foo1
+bar1
+```
+
+## Watch historical changes of keys
+
+Applications may want to watch for historical changes of keys in etcd for various reasons. For example, it is not unusual that an application wants to receive all the modifications of a key. If the application always keep connection to etcd, then `watch` is good enough. However, application or the etcd may fail. If there is a change happens during the failure, the application will not receive the update in real time. To get the update, application needs to be able to watch the historical changes of keys. To do this, application can specify a historical revision just like reading past version of keys.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Applications may want to watch for historical changes of keys in etcd. For example, an application may wish to receive all the modifications of a key; if the application stays connected to etcd, then `watch` is good enough. However, if the application or etcd may fails, a change may happen during the failure, and the application will not receive the update in real time. To guarantee the update is delivered, the application must be able to watch for historical changes to keys. To do this, an application can specify a historical revision on a watch, just like reading past version of keys.

File: Documentation/interact_v3.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+Here is the command to delete keys ranging from `foo` to `foo10`:
+
+```bash
+$ etcdctl del foo foo10
+2 # two keys are deleted
+```
+
+## Watch changes of keys
+
+Applications can watch on a key or a range of keys to get any related modification updates.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Applications can watch on a key or a range of keys to monitor for any updates.

File: Documentation/interact_v3.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+bar
+foo1
+bar1
+foo3
+bar3
+```
+
+## Read past version of keys
+
+Applications might want to read past version of a key for various reasons. For example, applications might want to rollback to an old configuration by getting an early version of a key or applications might want to get a consistent view over multiple keys over multiple requests. Every modification of the key changes the global revision of etcd cluster. To read an old version of key, application can specify a historical revision.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Applications may want to read superseded versions of a key for various reasons. For example,  an application may wish to rollback to an old configuration by accessing an early version of a key. Alternatively, an application may want a consistent view over multiple keys over multiple requests by accessing key history.
Since every modification to the etcd cluster key-value store increments the global revision of etcd cluster, an application can read superseded keys by providing a historical etcd revision.

File: Documentation/interact_v3.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+``` bash
+$ etcdctl put foo bar
+OK
+```
+
+## Read keys
+
+Applications can read values of keys from etcd cluster. Both range query and single key query are supported. 
+
+Suppose we have keys already stored in etcd as blow:
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Suppose an etcd cluster already has the following keys:

File: Documentation/interact_v3.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+Here is the command to set the value of key `foo` to `bar`:
+
+``` bash
+$ etcdctl put foo bar
+OK
+```
+
+## Read keys
+
+Applications can read values of keys from etcd cluster. Both range query and single key query are supported.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Both ranged queries and single-key queries are supported.

File: Documentation/interact_v3.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+By default, etcdctl talks to etcd server with v2 API for backward compatibility reason. To make etcdctl talks to etcd with v3 API, we first need to set API version to 3 by setting environment. 
+
+``` bash
+export ETCDCTL_API=3
+```
+
+## Write a key
+
+Applications can store keys into etcd cluster. The stored key will be replicated to all etcd members in the cluster through raft to achieve consistency and reliability.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: Applications store keys into the etcd cluster by writing to keys. Every stored key is replicated to all etcd cluster members through raft to achieve consistency and reliability.

File: Documentation/interact_v3.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
+# Interact with etcd
+
+Users mostly interact with etcd by putting or getting the value of a key. This section describes how to do that by using etcdctl, a command line tool for interacting with etcd server. The concepts described here should apply to the gRPC APIs or client library APIs.
+
+By default, etcdctl talks to etcd server with v2 API for backward compatibility reason. To make etcdctl talks to etcd with v3 API, we first need to set API version to 3 by setting environment.
```
Comment (heyitsanthony) [inline Documentation/interact_v3.md]: By default, etcdctl talks to the etcd server with the v2 API for backward compatibility. For etcdctl to speak to etcd using the v3 API, the API version must be set to version 3 via the `ETCDCTL_API` environment variable.

**Comment by xiang90** [comment]
@heyitsanthony PTAL

File: Documentation/interacting_v3.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+By default, etcdctl talks to the etcd server with the v2 API for backward compatibility. For etcdctl to speak to etcd using the v3 API, the API version must be set to version 3 via the `ETCDCTL_API` environment variable.
+
+``` bash
+export ETCDCTL_API=3
+```
+
+## Write a key
+
+Applications store keys into the etcd cluster by writing to keys. Every stored key is replicated to all etcd cluster members through raft to achieve consistency and reliability.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/raft/the Raft protocol/

File: Documentation/interacting_v3.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+Here is the command to set the value of key `foo` to `bar`:
+
+``` bash
+$ etcdctl put foo bar
+OK
+```
+
+## Read keys
+
+Applications can read values of keys from etcd cluster. Both ranged queries and single-key queries are supported.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/etcd cluster/an etcd cluster/

Comment (joshix) [inline Documentation/interacting_v3.md]: Replace second sentence:
"Queries may read a single key, or a range of keys."

File: Documentation/interacting_v3.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+``` bash
+$ etcdctl put foo bar
+OK
+```
+
+## Read keys
+
+Applications can read values of keys from etcd cluster. Both ranged queries and single-key queries are supported. 
+
+Suppose we have keys already stored in etcd as blow:
```
Comment (joshix) [inline Documentation/interacting_v3.md]: "Suppose the etcd cluster has stored the following keys:"

File: Documentation/interacting_v3.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+bar
+foo1
+bar1
+foo3
+bar3
+```
+
+## Read past version of keys
+
+Applications may want to read superseded versions of a key for various reasons. For example, an application may wish to rollback to an old configuration by accessing an early version of a key. Alternatively, an application may want a consistent view over multiple keys over multiple requests by accessing key history.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/for various reasons//
s/rollback/roll back/
s/early/earlier/
s/over multiple requests/through multiple requests/

File: Documentation/interacting_v3.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+Since every modification to the etcd cluster key-value store increments the global revision of etcd cluster, an application can read superseded keys by providing a historical etcd revision.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/of etcd cluster/of an etcd cluster/
s/a historical/an older/

File: Documentation/interacting_v3.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+$ etcdctl get --rev=2 foo foo10 # access the versions of keys at revision 2
+foo
+bar
+
+$ etcdctl get --rev=1 foo foo10 # access the versions of keys at revision 1
+```
+
+## Delete keys
+
+Applications can delete a key or a range of keys from etcd cluster.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/etcd/an etcd/

File: Documentation/interacting_v3.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+Here is the command to delete key `foo`:
+
+```bash
+$ etcdctl del foo
+1 # one key is deleted
+```
+
+Here is the command to delete keys ranging from `foo` to `foo10`:
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/`foo10`/`foo9`/

File: Documentation/interacting_v3.md [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+```bash
+$ etcdctl del foo foo10
+2 # two keys are deleted
+```
+
+## Watch changes of keys
```
Comment (joshix) [inline Documentation/interacting_v3.md]: `## Watch key changes`

File: Documentation/interacting_v3.md [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+Here is the command to watch on key `foo`:
+
+```bash
+$ etcdctl watch foo 
+# in another terminal: etcdctl put foo bar
+foo
+bar
+```
+
+Here is the command to watch on a range key from `foo` to `foo10`:
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/foo10/foo9/

File: Documentation/interacting_v3.md [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+```bash
+$ etcdctl watch foo foo10
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/foo10/foo9/

File: Documentation/interacting_v3.md [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+foo
+bar
+# in another terminal: etcdctl put foo1 bar1
+foo1
+bar1
+```
+
+## Watch historical changes of keys
+
+Applications may want to watch for historical changes of keys in etcd. For example, an application may wish to receive all the modifications of a key; if the application stays connected to etcd, then `watch` is good enough. However, if the application or etcd may fails, a change may happen during the failure, and the application will not receive the update in real time. To guarantee the update is delivered, the application must be able to watch for historical changes to keys. To do this, an application can specify a historical revision on a watch, just like reading past version of keys.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/may fails/fails/

File: Documentation/interacting_v3.md [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+# watch for changes on key `foo` since revision 3
+$ etcdctl watch --rev=3 foo
+PUT
+foo
+bar_new
+```
+
+## Compacted revisions
+
+As we mentioned, etcd keeps revisions so that applications can read past versions of keys. However, to avoid accumulating an unbounded amount of history, it is important to compact away past revisions. After a compaction, etcd removes historical revisions, releasing resources for future use. All superseded data with revisions before the compacted revision will be unavailable.
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/compact away/compact/
s/After a compaction/After compacting/

File: Documentation/interacting_v3.md [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+
+Here is the command to compact the revisions:
+
+```bash
+$ etcdctl compact 5
+compacted revision 5
+
+# any revisions before the compacted one are not be accessible
```
Comment (joshix) [inline Documentation/interacting_v3.md]: s/not be/not/

File: Documentation/interacting_v3.md (hunk 17 above)
Comment (xiang90) [inline Documentation/interacting_v3.md]: why this change? can you explain more?

**Comment by xiang90** [comment]
@joshix All fixed. (changed all foo10 to foo9 for readability)

File: Documentation/interacting_v3.md (hunk 17 above)
Comment (joshix) [inline Documentation/interacting_v3.md]: It is just for consistency across the examples - the first range example used `foo` through `foo9`.

**Comment by joshix** [comment]
LGTM
